apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,322 @@
+            });
+          });
+        }.bind(this);
+        break;
+    }
+  },
+
+  /**
+   * Dispatch events to bluetooth_transfer.js
+   * while getting bluetooth file transfer start/complete system messages
+   *
+   * @private
+   * @param  {String} msg message
+   * @param  {Object} transferInfo relayed info
+   */
+  _relayMessageEvent: function bt__relayMessageEvent(msg, transferInfo) {
+    this._setProfileConnected(this.Profiles.OPP, true);
+    var evt = document.createEvent('CustomEvent');
+    evt.initCustomEvent(msg,
+      /* canBubble */ true, /* cancelable */ false,
+      {transferInfo: transferInfo});
+    window.dispatchEvent(evt);
+  },
+
+  /**
+   * Get adapter for BluetoothTransfer when everytime bluetooth is enabled.
+   *
+   * @private
+   */
+  _initDefaultAdapter: function bt_initDefaultAdapter() {
+    this.debug('init bluetooth adapter');
+    this.getAdapter().then((adapter) => {
+      this._updateProfileStat(adapter);
+    });
+  },
+
+  /**
+   * Maintain connect stat of supported profiles.
+   *
+   * @private
+   * @param  {Object} adapter bluetooth adapter
+   */
+  _updateProfileStat: function bt_updateProfileStat(adapter) {
+    /* for v1, we only support two use cases for bluetooth connection:
+     *   1. connecting with a headset
+     *   2. transfering a file to/from another device
+     * So we need to listen to corresponding events to know we are (aren't)
+     * connected, then summarize to an event and dispatch to StatusBar
+     */
+    // In headset connected case:
+    adapter.addEventListener('hfpstatuschanged', (evt) => {
+        this._setProfileConnected(this.Profiles.HFP, evt.status);
+    });
+
+    adapter.addEventListener('a2dpstatuschanged', (evt) => {
+      this._setProfileConnected(this.Profiles.A2DP, evt.status);
+    });
+
+    adapter.addEventListener('scostatuschanged', (evt) => {
+      this._setProfileConnected(this.Profiles.SCO, evt.status);
+    });
+  },
+
+  /**
+   * Update bluetooth adapter.
+   *
+   * @private
+   * @param  {Object} resolve promise resolve function
+   * @param  {Object} evt event object
+   */
+  _bluetoothAttrChangeHandler: function bt__bluetoothAttrChangeHandler(resolve, evt) {
+    for (var i in evt.attrs) {
+      switch (evt.attrs[i]) {
+        case 'defaultAdapter':
+          this._adapter = this._bluetooth.defaultAdapter;
+          this.debug('defaultAdapter changed.');
+          resolve(this._adapter);
+        default:
+          break;
+      }
+    }
+  },
+
+  /**
+   * Called by external for re-use adapter.
+   *
+   * @public
+   */
+  getAdapter: function bt_getAdapter() {
+    return new Promise(function(resolve, reject) {
+      if (!this._bluetooth.defaultAdapter) {
+        this.debug('try to get adapter');
+        // need time to get bluetooth adapter at first run
+        this._bluetooth.addEventListener('attributechanged',
+          this._bluetoothAttrChangeHandler.bind(this, resolve));
```

Comment:
Be careful time issue. It's better to add event listener before check `defaultAdapter` is existed or not. We can add event listener first. And remove the listener if we don't want to care about `attributechanged` changed from the default adapter.